src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregationManager.cs

```diff
@@ -0,0 +1,274 @@
+            if (_collectTask != null || _cts.IsCancellationRequested)
+            {
+                // correct usage from internal code should never get here
+                Debug.Assert(false);
+                throw new InvalidOperationException("Start can only be called once");
+            }
+
+            if (_collectionPeriod.TotalSeconds < 1)
+            {
+                throw new InvalidOperationException("CollectionPeriod must be >= 1 sec");
+            }
+
+            var token = _cts.Token;
+            _collectTask = Task.Run(async () =>
+            {
+                double collectionIntervalSecs = -1;
+                lock (this)
+                {
+                    collectionIntervalSecs = _collectionPeriod.TotalSeconds;
+                }
+                if (collectionIntervalSecs < 1)
+                {
+                    Debug.Fail("_collectionPeriod must be >= 1 sec");
+                    return;
+                }
+
+                DateTime startTime = DateTime.UtcNow;
+                DateTime intervalStartTime = startTime;
+                while (!token.IsCancellationRequested)
+                {
+                    // intervals end at startTime + X*collectionIntervalSecs. Under normal
+                    // circumstance X increases by 1 each interval, but if the time it
+                    // takes to do collection is very large then we might need to skip
+                    // ahead multiple intervals to catch back up. Find the next interval
+                    // start time that is still in the future.
+                    DateTime now = DateTime.UtcNow;
+                    double secsSinceStart = (now - startTime).TotalSeconds;
+                    double alignUpSecsSinceStart = Math.Ceiling(secsSinceStart / collectionIntervalSecs) *
+                        collectionIntervalSecs;
+                    DateTime nextIntervalStartTime = startTime.AddSeconds(alignUpSecsSinceStart);
+
+                    // pause until the interval is complete
+                    try
+                    {
+                        TimeSpan delayTime = nextIntervalStartTime - now;
+                        await Task.Delay(delayTime, token).ConfigureAwait(false);
+                    }
+                    catch (TaskCanceledException)
+                    {
+                        return;
+                    }
+
+                    // collect statistics for the completed interval
+                    _beginCollection(intervalStartTime);
+                    Collect();
+                    _endCollection(intervalStartTime);
+                    intervalStartTime = nextIntervalStartTime;
+                }
+            });
+
+            _listener.Start();
+            _initialInstrumentEnumerationComplete();
+        }
+
+        public void Dispose()
+        {
+            _cts.Cancel();
+            if (_collectTask != null)
+            {
+                _collectTask.Wait();
```

Comment:
```suggestion
                _collectTask.GetAwaiter().GetResult();
```
? (to avoid the AggregateException).

Is DisposeAsync an option here?

Reply:
This got changed to thread.Join() as part of a switch from Task to Thread